docs/calyx-py.md

```diff
@@ -1,17 +1,120 @@
 # Emitting Calyx from Python
 
-Our frontends are written in Python3 and make use of the `calyx` library to
+Our frontends are written in Python3 and make use of the `calyx` builder library to
 generate their code.
 
+## Installation
+
 To install the library, run the following from the repository root (requires
 [flit][] installation):
+
 ```
 cd calyx-py && flit install -s
 ```
 
-The library provides an example:
+## Using the `calyx` builder
+
+The `calyx` library provides a builder to generate Calyx code. Let's walk through the file `calyx-py/test/builder_example.py`. This Calyx program initializes two registers with the numbers 1 and 41, adds them together, and stores the result in a register.
+
+We use the `Builder` object to construct our Calyx program `prog`. We can define components with `prog.component`. Here's a defininition of a `main` component with a 32-bit input and output:
+
+```python
+{{#include ../calyx-py/test/builder_example.py:init}}
 ```
-python calyx-py/test/example.py
+
+Technically, we didn't need to assign `prog.component("main")` to a variable; the `main` component would have been added to `prog` regardless. However, it will often prove useful to store handles to components, registers, or other objects you'd like to use later.
+
+We then instantiate our cells: three 32-bit registers and one 32-bit adder.
+
+```python
+{{#include ../calyx-py/test/builder_example.py:cells}}
+```
+
+As with adding components to a program, we don't need to assign `main.reg(...)` to a variable, but it'll be useful to be able to quickly refer to these cells.
+
+Next, we'll define our groups of assignments. The syntax for defining a group looks like `with {component}.group("group_name") as group_variable`, as we do below:
```

Comment:
I think we shouldn't split up the various parts of the group here. Just present the whole group first and then explain what the lines are doing.

Reply:
ok, i'll let you make this change if you want-- i found it pretty helpful to break it up and thought it was a little cumbersome without doing do. up to you though!